### PR TITLE
Solve breaking changes update training code

### LINKED ISSUE #4: Make activation function and the model optimizer configurable for MLP-like module
Currently, activation function (ReLU) and the optimizer (SGD) are hard-coded in the `MLPLitModule`. Making it configurable can help in creating experiments where the effects of the activation and optimizer can be studied/

### PR DESCRIPTION
**Pull request by RafayAK**
## What does this PR do?

Big PR 
- fixes breaking changes, 
- adds early stopping 
- learning rate scheduler
- the ability to set `max_epochs=-1` i.e infinity for overfitting
- breaks out optimizer initiation to configs for easily swapping out optimizers and their parameters  
- adds features to plot pretty decision boundaries.  

Fixes #4 #9

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?

No, this PR does multiple things as breaking changes required multiple updates to the codebase

- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?
☠☠☠☠☠
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by RafayAK** [comment]
For evaluation, here is what training results looked like when explicit overfitting, learning rate scheduler, and early stopping changes were not implemented:

image

https://wandb.ai/alex_data_augmentation/pinwheel_experiment_uniform_sampling_oracle_aug/runs/1dewwmsg


After changes:

image

https://wandb.ai/alex_data_augmentation/template-tests/runs/083vonau?

**Comment by RafayAK** [comment]
DO NOT MERGE these changes, it seems like GitHub action is broken, as automated tests are not running.
Let me figure that out.

**Comment by RafayAK** [comment]
Converting to draft, for now
